packages/react/src/components/ComposedModal/ComposedModal.tsx

```diff
@@ -0,0 +1,457 @@
+import React, {
+  useRef,
+  useEffect,
+  useState,
+  type MouseEvent,
+  type KeyboardEvent,
+  type HTMLAttributes,
+  type ReactNode,
+  type ReactElement,
+} from 'react';
+import { isElement } from 'react-is';
+import PropTypes from 'prop-types';
+import { ModalHeader, type ModalHeaderProps } from './ModalHeader';
+import { ModalFooter, type ModalFooterProps } from './ModalFooter';
+
+import cx from 'classnames';
+
+import toggleClass from '../../tools/toggleClass';
+import requiredIfGivenPropIsTruthy from '../../prop-types/requiredIfGivenPropIsTruthy';
+
+import wrapFocus from '../../internal/wrapFocus';
+import { usePrefix } from '../../internal/usePrefix';
+import { keys, match } from '../../internal/keyboard';
+
+export interface ModalBodyProps extends HTMLAttributes<HTMLDivElement> {
+  /** Specify the content to be placed in the ModalBody. */
+  children?: ReactNode;
+
+  /**
+   * Provide whether the modal content has a form element.
+   * If `true` is used here, non-form child content should have `bx--modal-content__regular-content` class.
+   */
+  hasForm?: boolean;
+
+  /**
+   * Specify whether the modal contains scrolling content
+   */
+  hasScrollingContent?: boolean;
+}
+
+export const ModalBody = React.forwardRef<HTMLDivElement, ModalBodyProps>(
+  function ModalBody(
+    {
+      className: customClassName,
+      children,
+      hasForm,
+      hasScrollingContent,
+      ...rest
+    },
+    ref
+  ) {
+    const prefix = usePrefix();
+    const contentClass = cx(
+      `${prefix}--modal-content`,
+      hasForm && `${prefix}--modal-content--with-form`,
+      hasScrollingContent && `${prefix}--modal-scroll-content`,
+      customClassName
+    );
+
+    const hasScrollingContentProps = hasScrollingContent
+      ? { tabIndex: 0, role: 'region' }
+      : {};
+
+    return (
+      <>
+        <div
+          className={contentClass}
+          {...hasScrollingContentProps}
+          {...rest}
+          ref={ref}>
+          {children}
+        </div>
+        {hasScrollingContent && (
+          <div className={`${prefix}--modal-content--overflow-indicator`} />
+        )}
+      </>
+    );
+  }
+);
+
+ModalBody.propTypes = {
+  /**
+   * Required props for the accessibility label of the header
+   */
+  // @ts-expect-error: Built-in prop-types > TS logic doesn't jive well with custom validators
+  ['aria-label']: requiredIfGivenPropIsTruthy(
+    'hasScrollingContent',
+    PropTypes.string
+  ),
```

Comment:
do we have a standard established for moving custom prop types to typescript? are we good to ignore the error? 🤔 @tay1orjones @mbarrer